### PR TITLE
Initial structure

### PR DESCRIPTION
**Pull request by jklukas**
We're currently using https://github.com/apakulov/kafka-graphite to send Kafka client metrics to Graphite, but that involves some hacky code to pull configuration out of the top-level `metrics` configuration of the application and duplicate it.

This project aims to provide more elegant integration with our Dropwizard services by registering Kafka client metrics to Dropwizard Metrics, thus using the same metrics pipeline as the rest of the application.

---

Once this is merged, I have some follow-up steps in mind:

- publish to Maven Central
- advertise to the Dropwizard community by submitting a PR to be included on http://modules.dropwizard.io/thirdparty/
- advertise to the Kafka community by asking to include a link on https://cwiki.apache.org/confluence/display/KAFKA/Ecosystem
- integrate into the Simple data team's `tricorder-kafka` library

### CODE REVIEW

**Comment by jklukas** [comment]
cc @kpfell @meli-lewis @xstevens @wrobstory; I don't know Allison's public GitHub username.

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>1.7</maven.compiler.source>
+    <maven.compiler.target>1.7</maven.compiler.target>
+  </properties>
+
+  <developers>
+      <developer>
+          <name>Jeff Klukas</name>
+          <email>jeff@klukas.net</email>
```
Comment (wrobstory) [inline pom.xml]: Just checking: do you want this email publicly associated with this project?

File: src/main/java/com/simple/metrics/kafka/DropwizardReporter.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package com.simple.metrics.kafka;
```
Comment (jklukas) [inline src/main/java/com/simple/metrics/kafka/DropwizardReporter.java]: Is this a reasonable package name? Is `kafka-dropwizard-reporter` a reasonable name for the module?

https://github.com/apakulov/kafka-graphite defines its reporter class in `org.apache.kafka.common.metrics` (the same package as the `MetricsReporter` interface that this implements) in order to get access to some package-private methods. Is that considered bad form to patch new classes into an existing package you don't own, or is that okay? Would it be more natural to users of this package to access the class as `org.apache.kafka.common.metrics.DropwizardReporter`?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,99 @@
...
+    @Override
+    public void run(MyConfiguration configuration, environment: Environment) {
+        SharedMetricRegistries.add("default", environment.metrics());
+    // [...]
+    }
+}
+```
+
+Metrics will be published with a prefix of
+`<metrics-prefix>.org.apache.kafka.common.metrics.MetricRegistry`
```
Comment (jklukas) [inline README.md]: We could shorten this to `<metrics-prefix>.org.apache.kafka` (which is what we've used internally), but it appears to be more Dropwizard-idiomatic to label these with the full classpath of the base class as you see here.

File: pom.xml (hunk 1 above)
Comment (jklukas) [inline pom.xml]: It's fine, I think? Does that seem not appropriate? Most of this pom is following closely [what's in jdub's pom](https://github.com/SimpleFinance/jdub/blob/master/pom.xml#L20-L31).

**Comment by xstevens** [comment]
+1

File: pom.xml (hunk 1 above)
Comment (wrobstory) [inline pom.xml]: It's really your personal preference regarding how comfortable you are with your email being out there for support emails.

File: README.md (hunk 3 above)
Comment (xstevens) [inline README.md]: I agree that classpath tends to be what is used. However I was just thinking it might be useful to prefix only to `<metrics-prefix>.org.apache.kafka.common.metrics`.

Comment (jklukas) [inline README.md]: I'm tempted by that shortcut, too. I was going to say that we should stick slavishly to the conventions of using full classpath in examples like https://github.com/dropwizard/metrics/blob/3.1-maintenance/metrics-logback/src/main/java/com/codahale/metrics/logback/InstrumentedAppender.java#L54 but I'm now realizing that our situation here is a bit different. In the InstrumentedAppender case, it's publishing metrics about what the base Appender class is doing. In our case, we're not publishing metrics about what MetricReporter is _doing_, but rather about metrics it's reporting.

Dropwizard's JVM metrics may be a better analogue, where they take metrics published as beans and republish them as Dropwizard metrics. In that case, the prefix is much shorter and doesn't represent a classpath.

If we're going to go with something that's not a full classpath, then I'd be for going all the way and using `<metrics-prefix>.org.apache.kafka`. I'd argue that the `common.metrics` piece is redundant; if we're republishing metrics, why care about where in the package structure the classes handling metrics are located?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,99 @@
...
+metric.reporters=com.simple.metrics.kafka.DropwizardReporter
+metric.dropwizard.registry=kafka-metrics
+```
+
+## Building
+
+To build the project, you'll need to
+[install Apache Maven 3](https://maven.apache.org/install.html).
+If you're on Mac OS X, installation via [Homebrew](http://brew.sh/)
+is recommended (`brew install maven`).
```
Comment (kpfell) [inline README.md]: I'd rephrase this slightly as 

> If you're on Mac OS X, you can install this via [Homebrew](http://brew.sh/):
> ```$ brew install maven```

or similar; having the command in parens at the end is a little odd to parse!

File: src/main/java/com/simple/metrics/kafka/DropwizardReporter.java (hunk 2 above)
Comment (kpfell) [inline src/main/java/com/simple/metrics/kafka/DropwizardReporter.java]: This seems like a reasonable package name, to me!

> Is that considered bad form to patch new classes into an existing package you don't own, or is that okay? Would it be more natural to users of this package to access the class as `org.apache.kafka.common.metrics.DropwizardReporter`?

Not sure whether it's "considered bad form," but I would find it potentially confusing and would give some lengthy side-eye. 😉

**Comment by kpfell** [comment]
+1

**Comment by jklukas** [comment]
Updated docs and going with @xavier's suggestion of `org.apache.kafka.common.metrics` as the prefix.
